plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -978,12 +992,35 @@ protected WriteMapping legacyToWriteMapping(@SuppressWarnings("unused") Connecto
                 .orElseGet(Function::identity);
     }
 
+    protected Function<String, String> tryApplyTopN(Optional<List<SortItem>> sortItems, OptionalLong limit)
+    {
+        if (sortItems.isEmpty() || limit.isEmpty()) {
+            return Function.identity();
+        }
+
+        return orderByFunction()
```

Comment:
Once again I must say I am too dumb to read a clever functional code like that (though I am sometimes not too dumb to write it :p ).

I would be more happy if instead

```
    @Override
    protected Optional<BiFunction<String, List<SortItem>, String>> orderByFunction()
    {
        return Optional.of((sql, sortItems) -> sql + " ORDER BY " + sortItems.stream()
                .map(this::getSortOrderSql)
                .collect(joining(", ")));
    }

```

we would have:

```
    @Override
    protected Optional<String> appendOrderBy(String sql, List<SortItem> sortItems)
    {
        String newSql = sql + " ORDER BY " + sortItems.stream()
                .map(this::getSortOrderSql)
                .collect(joining(", "));
        return Optional.of(newSql);
    }

```

Is there a huge benefit of using function objects instead coding logic in methods?

Reply:
I was trying to be consistent with existing code concepts.